File: Recursion/knapsack.py
#python3 syntax changes from Solution(object) or self class

def solve_knapsack(profits, weights, capacity):
  return knapsack_recursive(profits, weights, capacity, 0)


def knapsack_recursive(profits, weights, capacity, currentIndex):
  #base case check
  if capacity <= 0 or currentIndex >= len(profits):
    return 0


  #recursive case, if element at current index exceeds capacity, then leave
  profit1 = 0
  if weights[currentIndex] <= capacity:
    profit1 = profits[currentIndex] + knapsack_recursive(
      profits, weights, capacity-weights[currentIndex], currentIndex+1)

  #recursive call after exclude element at current index
  profit2 = knapsack_recursive(profits, weights, capacity,currentIndex+1)

  return max(profit1, profit2)

def main():
  print(solve_knapsack([1,6,10,16], [1,2,3,5], 7))
  print(solve_knapsack([1,6,10,16],[1,2,3,5],6))


main()

File: Recursion/knapsack.js
//Brute force recursion asymptotic O(2^n)
const solveKnapsack = (profits, weights, capacity) => {
  //having a closure inner function scope
  const knapsackRecursive=(profits, weights, capacity, currentIndex)=>{
    if(capacity <= 0 || currentIndex >=profits.length) return 0;
    let profit1=0;
    if(weights[currentIndex] <=capacity){
      profit1=profits[currentIndex] + knapsackRecursive(profits, weights, capacity-weights[currentIndex], currentIndex+1);
    }
    const profit2= knapsackRecursive(profits, weights, capacity, currentIndex+1);
    return Math.max(profit1, profit2);
  }
  return knapsackRecursive(profits, weights, capacity, 0);
};

const profits =[1,6,10,16];
const weights = [1,2,3,5];
console.log(`Total knapsack profit:--> ${solveKnapsack(profits, weights, 7)}`);
console.log(`Total knapsack profit: --> ${solveKnapsack(profits, weights, 6)}`);